Docs: Update migration guide for 8.0

kylegach opened the pull request:
## What I did

Updated the official migration guide for `8.0`

## Checklist for Contributors

### Testing

#### Manual testing

1. Follow the steps in the [contributing instructions](https://storybook.js.org/docs/react/contribute/new-snippets#preview-your-work) for this branch, `docs-migration-8`

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

kylegach committed:
Update migration guide for 8.0

jonniebigodes commented on docs/migration-guide.md:
````diff
@@ -58,171 +82,62 @@ To add Storybook to a project that isn’t currently using Storybook:
 
-#### Automatically detect MDX2 errors with a CLI tool
+```sh
+# Convert storiesOf to CSF
````
@kylegach can you remove the comments in the shell commands or else if a user copies the snippet it will also copy it as well and if they paste it in a terminal it can cause issues.

kylegach replied:
As of the docs refresh in November, the first line comment is now pulled out of the snippet and used as a title. So clicking the copy button only copies the command itself for this.

![](https://github.com/storybookjs/storybook/assets/486540/a14ab27c-7624-4e44-a0ae-ef88395839fe)

jonniebigodes commented on docs/migration-guide.md:
@kylegach we could potentially have this in a `Callout` component and push it upwards after the code snippet

kylegach replied:
The Callout draws a bit _too_ much attention to such a minor detail, imo.

jonniebigodes commented on docs/migration-guide.md:
@kylegach what do you think about adding a note about the [`doctor`](https://storybook.js.org/blog/storybook-7-6/) command here to help with the migration, which we can link to once it's added into the documentation? Just a thought, and feel free to disregard it if not applicable.

kylegach replied:
That's a great idea!

jonniebigodes reviewed (commented):
@kylegach aware that this is a draft pr which will be updated as we go, but left a couple of items for you to look into when able.

kylegach committed:
Reference `doctor` command

chantastic reviewed (approved):
This is a really nice doc. Well written and thoroughly referenced (with more specific links) ✅

jonniebigodes commented on docs/migration-guide.md:
One thing: when you're able, replace this with a code snippet aimed at every package manager. We have it, and it's called `storybook-migration-mdx-stories.npm.mdx` and include the other package manager variants. The same pattern should be applied to the examples here, as we should provide what's currently supported and not have users tied to a specific command, leaving up to them to use what they feel comfortable with.

kylegach replied:
@jonniebigodes — There is no snippet with that file name. There's also no harm in only providing the `npx` version of these codemod examples, as anyone with `node` installed already has `npx` available and none of these codemods relate to the project's package manager.

jonniebigodes replied:
@kylegach we have it it's [here](https://github.com/storybookjs/storybook/blob/next/docs/snippets/common/storybook-migration-mdx-stories.npm.mdx), and the pnpm version is [here](https://github.com/storybookjs/storybook/blob/next/docs/snippets/common/storybook-migration-mdx-stories.pnpm.mdx) and the Yarn version is [here](https://github.com/storybookjs/storybook/blob/next/docs/snippets/common/storybook-migration-mdx-stories.yarn.mdx).

kylegach replied:
Thanks. It must've been added recently and I hadn't pulled down the latest `next` into this branch. 🤦‍♂️

yannbf commented on docs/migration-guide.md:
```suggestion
**Note:** this migration supports the Storybook 6 ["CSF stories with MDX docs"](https://github.com/storybookjs/storybook/blob/6e19f0fe426d58f0f7981a42c3d0b0384fab49b1/code/addons/docs/docs/recipes.md#csf-stories-with-mdx-docs) recipe.
```

yannbf commented on docs/migration-guide.md:
```suggestion
Storybook now requires that MDX pages reference stories written in CSF, rather than the previous .stories.mdx hybrid approach. You can automatically convert your files using the following codemod (make sure to update the glob to fit your files):
```

yannbf commented on docs/migration-guide.md:
the only package which doesn't fall into the `@storybook` namespace which people should never remove is `storybook`, which powers the CLI!

yannbf commented on docs/migration-guide.md:
Did you mean to use `7.0.0-alpha.0` in this explanation? Or was it more like `8.0.0-alpha.0`?

yannbf commented on docs/migration-guide.md:
Reminder comment so we don't merge the PR with this

shilman replied:
I'd include the CSF2 to 3 migration in this list

yannbf commented:
I foresee people migrating from Storybook 6 to 8 (seems like [over 30%](https://craigory.dev/npm-burst/?package=@storybook/core-events) of the npm downloads are v6). So what does this PR mean for people who are looking for instructions from 6 to 7? Will they figure out that they need to switch versions in the docs so they get a sidebar entry for the Storybook 7 migration?

shilman commented on docs/migration-guide.md:
```diff
@@ -1,21 +1,45 @@
+- [Addons API introduced in Storybook 7 is now required](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#new-addons-api)
 
-If any of these apply to your project, please read through the [full migration notes](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#from-version-65x-to-700) before continuing. If any of these new requirements or changes do not fit your project, you should probably stick with Storybook 6.x.
+If any of these apply to your project, please read through the [full migration notes](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#from-version-7x-to-800) before continuing. If any of these new requirements or changes do not fit your project, you should probably stick with Storybook 7.x.
```
I'm not sure why these changes necessitate a full read of the migration notes. I'd say that if any of these apply to your project, you should definitely read those sections in detail.

And everybody might benefit from reading the detailed notes before upgrading OR can try to upgrade and consult the detailed notes as they troubleshoot problems.

kylegach replied:
Fair point!

shilman commented on docs/migration-guide.md:
```diff
@@ -58,171 +82,63 @@ To add Storybook to a project that isn’t currently using Storybook:
-We published a [detailed post about CSF3](https://storybook.js.org/blog/storybook-csf3-is-here/) including all of its benefits and migration steps.
-
-If you want to just skip to the migration, we provide a codemod for your convenience which should automatically make the code changes for you (make sure to update the glob to fit your files):
+Then, you can optionally convert CSF 2 to CSF 3 ([why should you do this?](https://storybook.js.org/blog/storybook-csf3-is-here/)):
```
Note that `storiesOf-to-CSF` converts to CSF1. There is no automated transformation from CSF 1 to 2.

kylegach replied:
@shilman — Here's what I came up with:

> ```sh
> # Convert storiesOf to CSF 1
> npx storybook@latest migrate storiesof-to-csf --glob="**/*.stories.tsx" --parser=tsx
> ```
> 
> While stories written in CSF 1 are supported in Storybook 8, we recommend upgrading to CSF 3, which has [significant benefits](https://storybook.js.org//blog/component-story-format-3-0/) when writing and maintaining stories. You'll need to manually [migrate your stories to CSF 2](https://storybook.js.org//blog/introducing-storybook-args/), and then you can use the [codemod to automatically upgrade from CSF 2 to 3](https://github.com/storybookjs/storybook/blob/docs-migration-8/docs/migration-guide.md#csf-2-to-csf-3).

I don't love it. 😕 But we don't have any materials that go straight from CSF 1 to 3, and providing a true guide for that (step-by-step, with multiple renderer snippets) would be a ton of effort for a small set of users that will encounter this. Wdyt?

shilman replied:
@kylegach I don't know if there's much value in converting from 1 to 2 and then 2 to 3.

How about something like this:

```
# Convert storiesOf to CSF 1
npx storybook@latest migrate storiesof-to-csf --glob="**/*.stories.tsx" --parser=tsx
```

This will create your stories to CSF1 stories, which are story functions that don't accept args. These are fully supported in Storybook 8, and will continue to be for the foreseeable future. However, we recommend writing all new stories in CSF3, which are story objects that are easier to write, reuse, and maintain.

kylegach replied:
My thinking was that CSF 1 (no args) to CSF 3 (implicit render functions, stories are objects, etc) would be quite a leap. So easier for the manual part to go to CSF 2, then let the automod handle it.

But I agree that it's weird.

-----

Oh, I just caught the "all _new_ stories" bit of your edit! That works for me. Thanks!

shilman commented on docs/migration-guide.md:
````diff
@@ -58,171 +82,63 @@ To add Storybook to a project that isn’t currently using Storybook:
 
 ```sh
-npx storybook@latest migrate csf-2-to-3 --glob="src/**/*.stories.js"
+# Convert CSF 2 to CSF 3
````
You should probably add this in another section, where you give a brief explanation of CSF2 and why you should upgrade. Note that this is an optional migration, since CSF2 is still fully supported (as is CSF1).

kylegach committed:
Apply suggestions from code review

Co-authored-by: Yann Braga <yannbf@gmail.com>

kylegach committed:
Address comments

- Clarify neccessity of reading migration notes in full vs. section
- Clarify that `storiesof-to-csf` codemod produces CSF 1
    - Add note about recommended upgrade to CSF 3
- Move CSF 2 to 3 upgrade to optional section
- Use existing snippets
- Fix typos
- Improve link URLs

kylegach committed:
Snippetize more code examples

- Update filenames to consistent convention

kylegach commented:
@yannbf —

> I foresee people migrating from Storybook 6 to 8 (seems like [over 30%](https://craigory.dev/npm-burst/?package=@storybook/core-events) of the npm downloads are v6). So what does this PR mean for people who are looking for instructions from 6 to 7? Will they figure out that they need to switch versions in the docs so they get a sidebar entry for the Storybook 7 migration?

That's what this section is for!

![](https://github.com/storybookjs/storybook/assets/486540/08100ab2-290a-445e-a074-60d6eaaf32c7)

That link is versioned. 😎

kylegach committed:
Clarify CSF 1 details
